core/client/fs/src/main/java/alluxio/client/file/UfsBaseFileSystem.java

```diff
@@ -0,0 +1,425 @@
+import alluxio.grpc.SetAclPOptions;
+import alluxio.grpc.SetAttributePOptions;
+import alluxio.grpc.UnmountPOptions;
+import alluxio.resource.CloseableResource;
+import alluxio.security.authorization.AclEntry;
+import alluxio.security.authorization.Mode;
+import alluxio.underfs.UfsFileStatus;
+import alluxio.underfs.UfsManager;
+import alluxio.underfs.UfsStatus;
+import alluxio.underfs.UnderFileSystem;
+import alluxio.underfs.options.CreateOptions;
+import alluxio.underfs.options.DeleteOptions;
+import alluxio.underfs.options.ListOptions;
+import alluxio.underfs.options.MkdirsOptions;
+import alluxio.util.CommonUtils;
+import alluxio.util.ModeUtils;
+import alluxio.util.io.PathUtils;
+import alluxio.wire.BlockLocationInfo;
+import alluxio.wire.FileInfo;
+import alluxio.wire.MountPointInfo;
+import alluxio.wire.SyncPointInfo;
+
+import com.google.common.base.Preconditions;
+import com.google.common.io.Closer;
+import io.grpc.Status;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Consumer;
+import java.util.stream.Collectors;
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * An implementation of the {@link FileSystem} interface that directly interacts with a target
+ * UFS. No Alluxio server is involved.
+ */
+@ThreadSafe
+public class UfsBaseFileSystem implements FileSystem {
+  private static final Logger LOG = LoggerFactory.getLogger(UfsBaseFileSystem.class);
+
+  /** Used to manage closeable resources. */
+  private final Closer mCloser = Closer.create();
+  protected final FileSystemContext mFsContext;
+  protected final CloseableResource<UnderFileSystem> mUfs;
+  protected final AlluxioURI mRootUFS;
+  protected volatile boolean mClosed = false;
+
+  /**
+   * Constructs a new base file system.
+   *
+   * @param fsContext file system context
+   */
+  public UfsBaseFileSystem(FileSystemContext fsContext) {
+    mFsContext = fsContext;
+    String ufsAddress = mFsContext.getClusterConf().getString(PropertyKey.USER_UFS_ADDRESS);
+    Preconditions.checkArgument(!ufsAddress.isEmpty(), "ufs address should not be empty");
+    mRootUFS = new AlluxioURI(ufsAddress);
+    UfsManager.UfsClient ufsClient = new UfsManager.UfsClient(
+        () -> UnderFileSystem.Factory.create(ufsAddress, mFsContext.getClusterConf()),
+        new AlluxioURI(ufsAddress));
+    mUfs = ufsClient.acquireUfsResource();
+    mCloser.register(mFsContext);
+    mCloser.register(mUfs);
+    LOG.debug("Creating file system connecting to ufs address {}", ufsAddress);
+  }
+
+  /**
+   * Shuts down the FileSystem. Closes all thread pools and resources used to perform operations. If
+   * any operations are called after closing the context the behavior is undefined.
+   */
+  @Override
+  public synchronized void close() throws IOException {
+    if (!mClosed) {
+      mClosed = true;
```

Comment:
should we add a return statement here?  I'm just thinking there might be an issue if we call mCloser.close() multiple times.